docs/usage/shoot_pod_autoscaling_best_practices.md

```diff
@@ -0,0 +1,327 @@
+- **Refine Thresholds:** If you notice frequent scaling actions or periods of underutilization or overutilization, adjust the thresholds used by HPA and VPA to better match the workload patterns.
+- **Update Policies:** Modify VPA update policies if you observe that the current settings are causing too much or too little pod disruption.
+- **Custom Metrics:** If using custom metrics, ensure they accurately reflect the load on your application and adjust them if they do not.
+- **Scaling Limits:** Review and adjust the minimum and maximum scaling limits to prevent over-scaling or under-scaling based on the capacity of your cluster and the criticality of your applications.
+
+# Quality of Service (QoS)
+
+A few words on the [quality of service for pods](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod). Basically, there are 3 classes of QoS and they influence the eviction of pods when `kube-reserved` is underrun, i.e. [node-level resources are running low](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#pod-selection-for-kubelet-eviction):
+- [`BestEffort`](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#besteffort), i.e. pods where no container has CPU or memory requests or limits: **Avoid them** unless you have really good reasons. The kube-scheduler will place them just anywhere according to its policy, e.g. `balanced` or `bin-packing`, but whatever resources these pods consume, may bring other pods into trouble or even the kubelet and the container runtime itself, if it happens very suddenly.
+- [`Burstable`](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#burstable), i.e. pods where at least one container has CPU or memory requests and at least one has no limits or limits that don't match the requests: **Prefer them** unless you have really good reasons for the other QoS classes. Always specify proper requests or use VPA to recommend those. This helps the kube-scheduler to make the right scheduling decisions. Not having limits will additionally provide upward resource flexibility, if the node is not under pressure.
+- [`Guaranteed`](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#guaranteed), i.e. pods where all containers have CPU and memory requests and equal limits: **Avoid them** unless you really know the limits or throttling/killing is intended. While `Guaranteed` sounds like something "positive" in the English language, this class comes with the downside, that pods will be actively CPU-throttled and will actively go OOM, even if the node is not under pressure and has excess capacity left. Worse, CPU throttling often goes unnoticed (VPA ignores throttling) and OOMs will not trigger evictions by the kubelet if the node-level resources are never running low (which will never happen, if all other pods together remain in their total requests; if you use VPA with proportional limit scaling, at least VPA may evict on OOMs).
```

Comment:
What I meant to say, though I agree that even I as the author took some time to reread and understand my own sentence (now after you raised the question):

Let's assume you use VPA with proportional limit scaling on a guaranteed pod, then the kubelet will (almost) never evict this pod, because it is guaranteed/gets its own cgroup (https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/ mentions that guaranteed pods would be evicted eventually based on their priority, but I really do not know how that can happen at all). So all depends on VPA then.

Then again, I agree, this sentence needs a rewrite. That's too much/too complicated. Let me try a rewrite...

Reply:
OK, the second part of the paragraph was indeed too convoluted and (actually) unnecessarily complex (why mention the kubelet at all if VPA will definitely notice, since this is a case that is only relevant with VPA enabled).

Thank you for the feedback. I changed the second part to:
> Worse, if containers in the pod are under VPA, their CPU requests/limits will often not be scaled up as CPU throttling will go unnoticed by VPA.